File: mil/sdk-ts/src/hash.ts
// MIL Hash64 identity derivations (design §3.2/§3.3), keyed BLAKE2b-512.
// Matches `misaka_mil_core::{ident, commit}`.

import { blake2b } from "@noble/hashes/blake2b";

const enc = new TextEncoder();

// Domain-separation strings (must equal the Rust `misaka-mil-v1/...` constants).
export const DOMAINS = {
  BIND: "misaka-mil-v1/bind",
  SESSION: "misaka-mil-v1/session",
  COMMIT: "misaka-mil-v1/commit",
  PROMPT_CT: "misaka-mil-v1/commit/prompt-ct",
  TRANSCRIPT: "misaka-mil-v1/transcript",
  PROVIDER_ID: "misaka-mil-v1/provider-id",
  QUOTE: "misaka-mil-v1/quote",
  KDF: "misaka-mil-v1/kdf",
  RECEIPT_CTX: "misaka-mil-v1/receipt/mldsa87",
} as const;

/** keyed BLAKE2b-512 (Hash64): key = domain (<=64 bytes), 64-byte output. */
export function hash64Keyed(domain: string, data: Uint8Array): Uint8Array {
  return blake2b(data, { key: enc.encode(domain), dkLen: 64 });
}

function concat(...parts: Uint8Array[]): Uint8Array {
  let n = 0;
  for (const p of parts) n += p.length;
  const out = new Uint8Array(n);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

/** report_data key binding: Hash64_k(bind, pk_kem ‖ pk_receipt). */
export function keyBinding(pkKem: Uint8Array, pkReceipt: Uint8Array): Uint8Array {
  return hash64Keyed(DOMAINS.BIND, concat(pkKem, pkReceipt));
}

/** provider id: Hash64_k(provider-id, pk_receipt). */
export function providerId(pkReceipt: Uint8Array): Uint8Array {
  return hash64Keyed(DOMAINS.PROVIDER_ID, pkReceipt);
}

/** session id: Hash64_k(session, quote_hash ‖ kem_ct ‖ nonce_req). */
export function sessionId(quoteHash: Uint8Array, kemCt: Uint8Array, nonceReq: Uint8Array): Uint8Array {
  return hash64Keyed(DOMAINS.SESSION, concat(quoteHash, kemCt, nonceReq));
}

/** inner prompt-ciphertext hash H(prompt_ct). */
export function promptCtHash(promptCt: Uint8Array): Uint8Array {
  return hash64Keyed(DOMAINS.PROMPT_CT, promptCt);
}

/** salted request commitment cm_req = Hash64_k(commit, salt ‖ H(prompt_ct)). */
export function requestCommitment(salt: Uint8Array, promptCtHashV: Uint8Array): Uint8Array {
  return hash64Keyed(DOMAINS.COMMIT, concat(salt, promptCtHashV));
}

export function requestCommitmentForCt(salt: Uint8Array, promptCt: Uint8Array): Uint8Array {
  return requestCommitment(salt, promptCtHash(promptCt));
}

/** Incremental response-transcript hasher producing cm_resp_k. The Rust hasher
 *  seeds the keyed BLAKE2b-512 state with the session id, then absorbs each
 *  response chunk; `commitment()` == blake2b(session_id ‖ chunks, key=TRANSCRIPT). */
export class TranscriptHasher {
  private absorbed: Uint8Array;
  constructor(sessionIdV: Uint8Array) {
    this.absorbed = new Uint8Array(sessionIdV); // seed prefix = session id
  }
  absorb(chunk: Uint8Array): void {
    const next = new Uint8Array(this.absorbed.length + chunk.length);
    next.set(this.absorbed);
    next.set(chunk, this.absorbed.length);
    this.absorbed = next;
  }
  commitment(): Uint8Array {
    return hash64Keyed(DOMAINS.TRANSCRIPT, this.absorbed);
  }
}

export function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i] ^ b[i];
  return diff === 0;
}

export function toHex(b: Uint8Array): string {
  let s = "";
  for (const x of b) s += x.toString(16).padStart(2, "0");
  return s;
}

export function fromHex(h: string): Uint8Array {
  const clean = h.startsWith("0x") ? h.slice(2) : h;
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(clean.substr(i * 2, 2), 16);
  return out;
}

File: mil/sdk-swift/Sources/MilSDK/Hash.swift
import Foundation

/// MIL Hash64 identity derivations (design §3.2/§3.3), keyed BLAKE2b-512.
/// Matches `misaka_mil_core::{ident, commit}` (cross-checked against Rust
/// vectors in the tests).
public enum MilHash {
    public enum Domain {
        public static let bind = Array("misaka-mil-v1/bind".utf8)
        public static let session = Array("misaka-mil-v1/session".utf8)
        public static let commit = Array("misaka-mil-v1/commit".utf8)
        public static let promptCt = Array("misaka-mil-v1/commit/prompt-ct".utf8)
        public static let transcript = Array("misaka-mil-v1/transcript".utf8)
        public static let providerId = Array("misaka-mil-v1/provider-id".utf8)
        public static let quote = Array("misaka-mil-v1/quote".utf8)
        public static let kdf = Array("misaka-mil-v1/kdf".utf8)
        public static let receiptCtx = Array("misaka-mil-v1/receipt/mldsa87".utf8)
    }

    public static func hash64Keyed(_ domain: [UInt8], _ data: [UInt8]) -> [UInt8] {
        Blake2b.keyed512(key: domain, data: data)
    }

    public static func keyBinding(pkKem: [UInt8], pkReceipt: [UInt8]) -> [UInt8] {
        hash64Keyed(Domain.bind, pkKem + pkReceipt)
    }

    public static func providerId(pkReceipt: [UInt8]) -> [UInt8] {
        hash64Keyed(Domain.providerId, pkReceipt)
    }

    public static func sessionId(quoteHash: [UInt8], kemCt: [UInt8], nonceReq: [UInt8]) -> [UInt8] {
        hash64Keyed(Domain.session, quoteHash + kemCt + nonceReq)
    }

    public static func promptCtHash(_ promptCt: [UInt8]) -> [UInt8] {
        hash64Keyed(Domain.promptCt, promptCt)
    }

    public static func requestCommitment(salt: [UInt8], promptCtHashV: [UInt8]) -> [UInt8] {
        hash64Keyed(Domain.commit, salt + promptCtHashV)
    }

    public static func requestCommitmentForCt(salt: [UInt8], promptCt: [UInt8]) -> [UInt8] {
        requestCommitment(salt: salt, promptCtHashV: promptCtHash(promptCt))
    }
}

/// Incremental response-transcript hasher producing cm_resp (design §3.3/§4.1):
/// seeded with the session id, absorbs each response chunk.
public struct TranscriptHasher {
    private var absorbed: [UInt8]
    public init(sessionId: [UInt8]) { absorbed = sessionId }
    public mutating func absorb(_ chunk: [UInt8]) { absorbed.append(contentsOf: chunk) }
    public func commitment() -> [UInt8] { MilHash.hash64Keyed(MilHash.Domain.transcript, absorbed) }
}

public enum Hex {
    public static func encode(_ bytes: [UInt8]) -> String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }
    public static func decode(_ s: String) -> [UInt8] {
        let clean = s.hasPrefix("0x") ? String(s.dropFirst(2)) : s
        var out = [UInt8]()
        var i = clean.startIndex
        while i < clean.endIndex {
            let j = clean.index(i, offsetBy: 2)
            out.append(UInt8(clean[i..<j], radix: 16)!)
            i = j
        }
        return out
    }
}
